### PR TITLE
HPCC-20896 Support ESP session cookie verification

### PR DESCRIPTION
**Pull request by wangkx**
An new ESP method /esp/verify is added to verify ESP session
cookies. The method can be used to verify a session ID
during the first load of any ECL Watch page. If no cookie
name is specified, all the cookies in HTTP Cookie header are
verified.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead please review.

File: esp/bindings/http/platform/httpservice.cpp [hunk 1]
```diff
@@ -1120,6 +1125,208 @@ EspAuthState CEspHttpServer::preCheckAuth(EspAuthRequest& authReq)
...
+        //a login process is started. At that time, any SESSION_START_URL_COOKIE should be invalid.
+        cookie.verificationDetails.set("ESP cannot verify this cookie. This cookie is only valid within a login process.");
+    }
+    else if (strieq(name, SESSION_ID_TEMP_COOKIE))
+    {   
+        //SESSION_ID_TEMP_COOKIE: used for remembering UserID when a user is in AS_PASSWORD_VALID_BUT_EXPIRED state and updating password.
+        //It is created before the updatepasswordinput is started. After the update process, ESP should remove this cookie if possible.
+        //This verify function is designed to verify cookies before a login process is started. At that time, any SESSION_ID_TEMP_COOKIE
+        //should be invalid.
+        cookie.verificationDetails.set("ESP cannot verify this cookie. This cookie is only valid when updating an expired passwsord.");
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: Misspelling (passwsord)

File: esp/bindings/http/platform/httpservice.cpp [hunk 2]
```diff
@@ -1120,6 +1125,208 @@ EspAuthState CEspHttpServer::preCheckAuth(EspAuthRequest& authReq)
...
+        ForEach(*iter)
+        {
+            const char* name = iter->getPropKey();
+            if (strieq(name, "__querystring"))
+                continue; //skip the property created by ESP.
+
+            //ESP client cannot access the session ID cookie (both cookie name and value).
+            //The client may use "ESPSessionID" to verify the session ID cookie.
+            if (strieq(name, "ESPSessionID"))
+                cookies.append(*(new CESPCookieVerification("ESPSessionID", "")));
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: Why is this necessary?  You already know "name" is "ESPSessionID", except perhaps for the casing of that string. Can't you always call
    cookies.append(*(new CESPCookieVerification(name, "")));

File: esp/bindings/http/platform/httpservice.cpp [hunk 3]
```diff
@@ -1120,6 +1125,208 @@ EspAuthState CEspHttpServer::preCheckAuth(EspAuthRequest& authReq)
...
+        cookie.cookieValue.set(m_request->queryCookie(USER_NAME_COOKIE));
+    if (cookie.cookieValue.get())
+        cookie.valid.set("true");
+}
+
+bool CEspHttpServer::verifyESPSessionIDCookie(EspAuthRequest& authReq)
+{
+    unsigned sessionID = readCookie(authReq.authBinding->querySessionIDCookieName());
+    if (sessionID == 0)
+        return false;
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: When could sessionID be 0?  No sessions enabled?  Perhaps a comment here would help

File: esp/bindings/http/platform/httpservice.cpp [hunk 4]
```diff
@@ -1120,6 +1125,208 @@ EspAuthState CEspHttpServer::preCheckAuth(EspAuthRequest& authReq)
...
+
+    //Timeout old sessions.
+    Owned<IRemoteConnection> conn = getSDSConnection(authReq.authBinding->queryESPSessionSDSPath(), RTM_LOCK_WRITE, SESSION_SDS_LOCK_TIMEOUT);
+    IPropertyTree* espSessions = conn->queryRoot();
+    if (authReq.authBinding->getServerSessionTimeoutSeconds() >= 0)
+    {
+        CDateTime now;
+        now.setNow();
+        time_t timeNow = now.getSimple();
+        if (timeNow - lastSessionCleanUpTime >= authReq.authBinding->getCheckSessionTimeoutSeconds())
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: I prefer to see parenthesis around "timeNow - lastSessionCleanUpTime"

File: esp/bindings/http/platform/httpservice.cpp [hunk 5]
```diff
@@ -1120,6 +1125,208 @@ EspAuthState CEspHttpServer::preCheckAuth(EspAuthRequest& authReq)
...
+            cookie.valid.set("true");
+        if (!checkValid && !m_request->queryCookie(USER_NAME_COOKIE))
+            cookie.valid.set("true");
+        return;
+    }
+
+    bool isSessionIDCookieValid = verifyESPSessionIDCookie(authReq);
+    if (checkValid && isSessionIDCookieValid)
+        cookie.valid.set("true");
+    if (!checkValid && !isSessionIDCookieValid)
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: I believe this could be coded as "else if"

**Review by RussWhitehead — CHANGES_REQUESTED** [review]
A few comments and questions

**Comment by wangkx** [comment]
@RussWhitehead please review my changes.

**Review by RussWhitehead — APPROVED** [review]
@wangkx Looks good

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

**Pull request merged**